### ISSUE TITLE
Control Name Clashes

### ISSUE THREAD

**Issue by telephon** [supercollider/supercollider#2832]
Currently, we may have clashing control names (this is known, and a suitable error is posted):

```
f = { |xxx| Control.names([\xxx]).kr(0).poll; 0.0 };
a = f.play;
a.set(\xxx, 100); // remains 0
```

The error is only posted when we use the function to build a synthdesc (`add`):
`SynthDef(\x, f).add;`

(actually the error is posted twice)

This is not so bad, one may think, because you know that you shouldn't be doing this in one function.

But this makes  one feature of `NamedControl` a little unstable, namely that you can refer to controls in a function passed to a function that builds a synth def:

```
// this is fine
g = { SinOsc.ar(\freq.kr(440)) };
f = { g.value * LFNoise2.kr(8, 0.2).max(0) };

x = f.play;
x.set(\freq, 1000);

// but not this:
f = { |freq| g.value * LFNoise2.kr(freq / 20, 0.2).max(0) };
x = f.play;
x.set(\freq, 1000);
```

The original intention for `NamedControl` had been this kind of modularity, however. (it will warn you if you use several NamedControls with different default values or rates).

It would be really good to get this straight, and it certainly could be possible. But the way controls are handled in SynthDefs is a bit complicated, and some important information is unavailable in the actual `Control` object that is returned.

Maybe someone has ideas what best to do? @muellmusik ?

**Comment by scztt** [comment]
I was looking at Control etc. recently - you're right, the implementation turns out to be pretty convoluted, and probably could benefit from some refactoring. The roles of `Control`, `NamedControl`, and `ControlName` are not particularly clear, and the naming doesn't help :). The relationship between a ControlName and Control seems to be maintained solely by their indexical value in the list of controls in a synth. And a NamedControl isn't really an object at all - it's new method (1) creates a ControlName, and (2) returns a Control linked to that ControlName - it behaves more like a special constructor for Control and it's subclasses.

I tried for an hour or so to clean up of the stuff around NamedControl, but it's a lot to untangle. One simple fix for the bug you mentioned is to add a check for existing ControlNames in SynthDef:addControlName. This should look like the checks that already happen in NamedControl - I tested this and it fixes a bunch of cases:

```
+Control {
	checkEquivalence {
		|other|

		if (other.defaultValue.notNil && defaultValue.notNil) {
			if(other.defaultValue.asArray != defaultValue.asArray) {
				^Error("Control: cannot have more than one set of "
					"default values in the same control.");
			}
		};

		if(other.rate.notNil && rate.notNil) {
			if (other.rate != rate) {
				^Error("Control: cannot have  more than one set of "
					"rates in the same control.");
			};
		};

		if (fixedLag) {
			if (other.lag.asArray != lag.asArray) {
				^Error("Control: Cannot have more than one set of "
					"lags if fixedLag==true");
			}
		};

		^nil; // These ControlNames are equivalent
	}
}

+SynthDef {
	addControlName { arg cn;
		var existing = allControlNames.detect({ |c| c.name == cn.name });

		if (existing.notNil) {
			var error = existing.checkEquivalence(existing);
			if (error.isNil) {
				cn = existing; // Is this the right thing to do here?
			} {
				error.throw;
			}
		};

		controlNames = controlNames.add(cn);
		allControlNames = allControlNames.add(cn);
	}
}
```

The next step would be to more or less factor NamedControl out entirely (maybe keep the class but redirect it for backwards compatibility). The functionality of NamedControl - creating a new ControlName linked to a Control, and then returning the Control - should probably live in a method in SynthDef, something like "createNamedControl". There's other code in SynthDef related to constructing a Control out of a ControlName, so there's probably some consolidation to be done there as well.

**Comment by scztt** [comment]
I've been experimenting with adding the ability with setting ControlSpec's directly on Controls, rather than having to add them indirectly via the SynthDef metadata (see https://gist.github.com/scztt/672e28a228cc34e3b2dd00dc25a4cb9d). This ends up improving readability of SynthDefs a lot - you can do:
```
SynthDef(\tone, {
   |freq|
   freq.controlName.spec = [20, 10000].asSpec;
   SinOsc.ar(freq);
});
```

The ability to set general properties of ControlNames after the fact seems useful as well - for one, it allows you to do some things with SynthDef function-arg-style parameters that you could formerly do only with NamedControls - for one, arrayed parameters. Hypothetically:

```
SynthDef(\tone, {
    |freq, env|
    freq.controlName.spec = [20, 10000].asSpec;
    freq.controlName.lag = 3;
    env.controlName.defaultValue = Env.adsr().asArray;
    SinOsc.ar(freq, 0, env);
});
```

**Comment by telephon** [comment]
@scztt this is excellent. 

I implemented `NamedControl` as an improvement in this direction, but your approach seems the better way. We need a way to access controls easily without access to the argument value. But that can then be:

`{ \freq.kr.controlName }`

To make `\symbol.kr` etc. work, we just need a method find the control ugen output for a given name.

And then it would also be possible to create a more tolerant adjustment schema - so if you request a control rate version of an already existing audio rate control, you'd get it wrapped in a `K2A`. This is a separate topic probably, what should warn and what not.

**Comment by vivid-synth** [comment]
Sorry for my ignorance, but is this entirely an issue with the class library? (I.e. this has nothing directly to do with the Control UGen?)

**Comment by scztt** [comment]
@vivid-synth Yes, it's related to how sclang builds synthdefs. Though it's definitely a potential bug in other client languages that construct SynthDefs - constructing synthdefs with duplicate param names is probably undefined and should be prevented by the client assembling the synthdef.
